### PR TITLE
Fix horrifying bug in lossless_cast of a subtract

### PR DESCRIPTION
**Pull request by abadams**
This is one thing breaking the SVE2 pr.

### CODE REVIEW

**Review by vksnk — APPROVED** [review]
Oof.

**Review by steven-johnson — APPROVED** [review]
owww

**Comment by zvookin** [comment]
Please sir, may I have a test?

**Comment by mcourteaux** [comment]
Hahahaha, I'm giggling for minutes now... :laughing:

**Comment by steven-johnson** [comment]
Is this a recent injection? If not, might want to backport to 17...

**Comment by mcourteaux** [comment]
> Is this a recent injection? If not, might want to backport to 17...

No, 4 years ago:
https://github.com/halide/Halide/commit/394536f5d7ffea57194294c8d5efbd805902ce88

**Comment by abadams** [comment]
I added a fuzz test, and there are more bugs :(

I'm pretty ready to state that all new features or algorithms must come with a fuzz tester.

**Comment by zvookin** [comment]
This is why people measure test coverage. It's kinda annoying, but probably a thing to consider.

**Comment by steven-johnson** [comment]
where do we stand on this -- do we need to do more work (tests, etc) before landing?

**Comment by abadams** [comment]
Still working on it. I couldn't figure out how to fix the remaining bugs using top-down reasoning about the types like what was there already, so I rewrote it to use bottom-up constant-integer bounds analysis, using code lifted out of the simplifier. It's working but there are inefficiencies to fix and some code deduplication to do.

**Comment by abadams** [comment]
lossless_negate has a similar bug. The mul visitor just tries to negate each arg, but if you have e.g. an int8 type, and the product is exactly -128 (e.g. -64 * 2), then a lossless negate just returns -128 again, which hasn't negated anything.

**Comment by steven-johnson** [comment]
Status Update? I know this had a lot of other stuff to fix first, just not sure where this stands.

**Comment by abadams** [comment]
The next PR is going to be reworking the simplifier to use the recently-merged ConstantInterval (pulling that out of this PR), and then after that's in this PR will be updated and hopefully made much smaller.

**Comment by mcourteaux** [comment]
Fun fact: just came across this bug while traversing the code trying to figure out what's wrong.

**Comment by abadams** [comment]
This is too much of a change to backport now

**Comment by mcourteaux** [comment]
> This is too much of a change to backport now

I don't know how important it is, but changing `+` to `-` as a backported hotfix seems easy.

**Comment by abadams** [comment]
Sure we could backport that. It was just the tip of the iceberg when it came to bugs in lossless_cast though, and fixing the other ones is what required all the work.

**Comment by abadams** [comment]
#8264

**Comment by abadams** [comment]
Tests are breaking because vulkan is busted on main, and because the fuzzer and asan builds are still busted on linux-worker-1. This is ready for review.

**Comment by steven-johnson** [comment]
> Tests are breaking because vulkan is busted on main, and because the fuzzer and asan builds are still busted on linux-worker-1. This is ready for review.

@derek-gerstmann is working on the Vulkan bug, but one of us should figure out the fuzzer/asan issue

**Review by steven-johnson — COMMENTED** [review]
LGTM but @vksnk should probably review too and he's out most of this week.

I'll do a full test of this inside Google later today.

File: src/CodeGen_X86.cpp [hunk 1]
```diff
@@ -736,7 +736,12 @@ void CodeGen_X86::visit(const Call *op) {
         // Handle edge case of possible overflow.
         // See https://github.com/halide/Halide/pull/7129/files#r1008331426
         // On AVX512 (and with enough lanes) we can use a mask register.
-        if (target.has_feature(Target::AVX512) && op->type.lanes() >= 32) {
+        ConstantInterval ca = constant_integer_bounds(a);
+        ConstantInterval cb = constant_integer_bounds(b);
+        if (!ca.contains(-32768) || !cb.contains(-32768)) {
```
Comment (mcourteaux) [inline src/CodeGen_X86.cpp:741]: `-0x8000` is probably clearer?

Comment (steven-johnson) [reply src/CodeGen_X86.cpp:741]: Anyone reading this code who doesn't know the bit layout of `-32768` should not be reading this code :-)

**Comment by abadams** [comment]
Is this now good to merge? I think it was contingent on the results of tests inside Google.

**Comment by steven-johnson** [comment]
Not yet -- need to retest in Google with the Xtensa codegen fix in place, but I'm OOO for at least another day or two (out of town due to family medical emergency. Can it wait till then?

**Comment by abadams** [comment]
Yes, there's no rush, I just wanted to make sure it's not stalled on me.

**Comment by vksnk** [comment]
Xtensa codegen is fixed now, so I re-run tests in Google codebase: there are a couple of tests which now produce mismatch with thegoldens (in one of the tests, the max diff is 6 out of 255, which seems fairly significant), but I guess this is somewhat expected?

**Comment by abadams** [comment]
Are the tests that differ using floats in some way that would be unstable? (e.g. using a floating-point comparison to decide on an interpolation direction in a demosaicking algorithm)

If they're fixed-point algorithms that's definitely not expected.

**Comment by vksnk** [comment]
It's hard to tell for sure, because it's one of those large end-to-end tests which can call 10 different Halide functions deep inside, but I'd guess it does use some floating-point math (I'm not 100%, but I think one of these tests was failing before due to some FP differences before, so we needed to regenerate the goldens; although, the diff was smaller I think).

**Comment by abadams** [comment]
Do you think it's worth testing if it passes on main under strict_float, and fails on this branch under strict_float? If that's the case there's probably something to worry about.

**Comment by vksnk** [comment]
Sure, I'll see if it can be narrowed down.

**Comment by steven-johnson** [comment]
Update on Google: it looks like there are no real red flags here -- just goldens to be regenerated -- so I'm hoping to be able to approve this for landing later today pending confirmation.

**Comment by steven-johnson** [comment]
Approved and landed in Google, so I'm landing this here now.
